### PR TITLE
Iframed editor: Fix relative wp-content URLs

### LINKED ISSUE #54262: Relative scheme images no longer display in the editor
### Description

In WordPress 6.3 and the current version of the Gutenberg plugin, 16.6.0, images with relative schemes are not displayed correctly in the block editor. On the front end they render as expected.

In the editor, the images are shown as a broken image and -- browser dependent -- the alt text shown in it's place.

### Step-by-step reproduction instructions

1. Install WordPress 6.3 locally
2. In the wp-config file, define a relative content URL constant: `define( 'WP_CONTENT_URL', '//example.local/wp-content' );`
3. Create a post in the block editor
4. Insert an image block
5. Upload an image to the image block/library
6. The image fails to load
7. Preview the post in a new tab: the image does display
8. Switch to the code editor
9. Modify the image URL to include the scheme
10. Exit the code editor
11. The image now displays

### Screenshots, screen recording, code snippet

**WordPress 6.3** (with and without the latest version of Gutenberg active)

Screen Shot 2023-09-08 at 10 48 42 am

**WordPress 6.2** (without Gutenberg active)

Screen Shot 2023-09-08 at 10 50 34 am

### Environment info

- WordPress 6.3
- Browsers: Reproduced in Firefox, Safari and Chrome (all latest versions)
- OS: macOS
- Virtual machine: https://github.com/chassis/chassis on VirtualBox
- Home URL: example.local
- Site (WP) URL: example.local/wp (ie WordPress is running in [its own sub-directory](https://wordpress.org/documentation/article/giving-wordpress-its-own-directory/))

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by Soean**
## What?
Adds a `<base>` element to the iframed editor to work with relative URLs.

Fixes #54262
Rebase of #56533

## Why?
Currently the editor in the iframe does not work with relative URLs.

Relative URLs work in the editor without iframe, where the media are loaded relative to the admin URL. The <base> element restores the old behavior. In further iterations, we can make the editor even more flexible so that it can also work with other URLs.

## Testing Instructions
- Make sure your WordPress uses only blocks with version v3 to that the iframed editor is used.
- Add the constant to your wp-config.php file: 
```php
define( 'WP_CONTENT_URL', '/wp-content' );
```
- Add a new image to a post. 
- The image should be previewed in the editor.

### CODE REVIEW

File: packages/block-editor/src/components/iframe/index.js [hunk 1]
```diff
@@ -272,6 +272,7 @@ function Iframe( {
 <html>
 	<head>
 		<meta charset="utf-8">
+		<base href="${ window.location.origin }">
```
Comment (youknowriad) [inline packages/block-editor/src/components/iframe/index.js:291]: Does this work, regardless of how the site is configured (top level, sub folder, ...) and also for third-party block editors?

Comment (ellatrix) [reply packages/block-editor/src/components/iframe/index.js:291]: This shouldn't matter for relative urls starting with `/`, they should always use the root.
But for relative urls that don't start with a slash it's trickier. We'd have always use the current post URL etc. This wouldn't work without an iframe either (in the admin), so not sure if it's worth fixing.

**Review by ellatrix — APPROVED** [review]
Sounds good!

**Comment by ellatrix** [comment]
Btw I just tested this manually setting an image block's URL.

**Comment by getdave** [comment]
I trust you'll understand that I've removed the `Backport` label from this PR 🙏 

This is because we are now past the final public RC and any further changes targeting the main 6.7 release need to be discussed and approved by a release lead. 

Looking at the PR it's unclear to me whether this is a bug introduced during the Beta/RC period or something that's been present for a while. If it's a longstanding bug then in the interests of stability I'd vote to defer this to 6.7.1.

**Update**: I noticed [the Issue refers to WordPress 6.3](https://github.com/WordPress/gutenberg/issues/54262) so this seems to be a longstanding bug.

Pinging @colorful-tones @ndiego @kevin940726 for confidence check as I may have missed something critical.

Also it looks like the PR needs a rebase against `trunk`.

**Comment by Soean** [comment]
@getdave As of WordPress 6.3, the Post Editor is iframed if all registered blocks have a Block API version 3 or higher and no traditional metaboxes are registered.

With WordPress 6.7 the editor is iframed also with metaboxes active:
https://make.wordpress.org/core/2024/10/18/post-editor-iframing-with-meta-boxes-in-wordpress-6-7/

So we have a problem after the update. Some of our sites are no longer editable in wp-admin. There is no filter or anything to disable it. We need to add a block with API V2 to get an editor that works with relative URLs again.

It's not a new bug, but it will now be visible to more users after the update to WordPress 6.7.

**Comment by ndiego** [comment]
Given the release's timing and the fact that this issue is not unique to 6.7, I would be inclined to punt to 6.7.1. 

> So we have a problem after the update. Some of our sites are no longer editable in wp-admin. There is no filter or anything to disable it. We need to add a block with API V2 to get an editor that works with relative URLs again.

That said, I agree that this will likely impact _many_ more people now that the Editor is iframed, even when metaboxes are active. To reduce the impact on users and developers, I recommend we try and include this in 6.7.

**Comment by getdave** [comment]
I'm afraid I didn't understand the impact of the original Issue until a few moments ago. Given the potential scope I'd also be inclined to include.

I see activity from @ellatrix and @youknowriad. How confident are you in this fix landing 6.7?

**Comment by ellatrix** [comment]
It would have been good if the impact was included in the PR or issue description.

Fwiw, I think this is safe. It will only affect _relative_ URLs in the block canvas of links and resources. WordPress doesn't even normally use relative links.

**Comment by ellatrix** [comment]
The failure is legit. This breaks the footnote back links, which are relative has urls. Since the `window.location` is `blob:http://...` and not `http://`, it will cause the iframe to be reloaded and show the root url because the base changes from the blob protocol to the http(s) protocol. So now the fix becomes more complex because we'll have to intercept all anchor tags to make sure they never reload the iframe. Not that this can't currently happen. In trunk, when you click a link in a block outside of rich text, it will also cause the page to reload, except for relative hash urls.

So either we fix all links, or we only fix relative hash links. In both cases we'd have to intercept clicks on anchors tags in the iframe.

**Comment by getdave** [comment]
This didn't make the RC4 which started now. 

I will raise this with the release team and we will make a call on whether it needs to be escalated. 

Thank you @ellatrix for finishing this up and @Soean for raising it.

**Comment by peterwilsoncc** [comment]
@getdave At this part of the release cycle, only bug fixes for regressions introduced in-cycle can be included so this will need to wait until WP 6.8.

**Comment by Soean** [comment]
With 6.6 it works. After the 6.7 update, all images in the backend are broken. I think that it should at least be included in 6.7.1.

But perhaps we should wait and see whether we are a special case or whether it is a problem for other people.

**Comment by getdave** [comment]
@peterwilsoncc Thanks for jumping in here. There's been some disagreement and discussion above about whether or not to try and include this, but yes this was an issue outside of the Beta / RC so on balance I think I stick by [my original analysis](https://github.com/WordPress/gutenberg/pull/66751#issuecomment-2461708388) - do _not_ include in 6.7.

I don't see any reason why we can't move to 6.7.1 though, unless there are some rules around what can be included in those point releases that I'm not aware of.

**Comment by Mamaduka** [comment]
I noticed a minor error with the fix - #68024.

**Comment by cbravobernal** [comment]
Punting this to 6.8. As it created another minor error and is starting to diverge quite a lot.

**Comment by Mamaduka** [comment]
@cbravobernal, added a minor release backport label back here. See: #68689.

**Comment by t-hamano** [comment]
I am checking the PRs with the "Backport to WP Minor Release" label applied to clarify which PRs should be backported to [the WordPress 6.8.2 release](https://make.wordpress.org/core/2025/06/17/wordpress-6-8-2-release-schedule/).

This PR is part of Gutenberg 19.2. Since the Gutenberg versions in WordPress 6.8 are from 19.4 to 20.4, this PR is already included in the 6.8 release. Therefore, there is no need to backport this PR to 6.8.2.

Removing the "Backport to WP Minor Release" label.

cc @audrasjb @estelaris @zunaid321

**Comment by jsnajdr** [comment]
@t-hamano @Mamaduka I found this old PR when debugging an issue where links in block UI are disabled. Jetpack has a "Blogging Prompt" block that displays a "View all responses" link:

Screenshot 2025-07-24 at 12 46 14

This PR, however, added a `click` listener that disables the links:
```js
iFrameDocument.addEventListener( 'click', ( event ) => {
  if ( event.target.tagName === 'A' ) {
    event.preventDefault();
    // Do some custom handling of the `href`
  }
} );
```
Therefore, clicking the link does nothing.

However, this doesn't seem to be the intent of the PR: it's about handling relative URLs in images etc.

Curiously, the Core Embed block also displays a link:

Screenshot 2025-07-24 at 12 53 14

But this one works because the markup is:
```
<a href="..."><span>Learn more about embeds</span></a>
```
and the `event.target.tagName` is `SPAN` when you click.

**Comment by Mamaduka** [comment]
Thanks for the ping, @jsnajdr!

It appears that @ellatrix introduced the link click interception - https://github.com/WordPress/gutenberg/pull/66751#issuecomment-2462468972. Maybe we can make it less restrictive.

Similar click "traps" are also not great for accessibility, and some blocks use special handling for them. Linking to some good discussions for the Latest Posts block:

* https://github.com/WordPress/gutenberg/pull/40593#pullrequestreview-952362910
* https://github.com/WordPress/gutenberg/pull/40593#issuecomment-1111295332

**Pull request merged**